File: Python/47. Permutations II.py
'''
横向遍历不重复数字，纵向递归不重复数字
'''
class Solution:
    def permuteUnique(self, nums: List[int]) -> List[List[int]]:
        path = []
        res = []
        used = [False]*len(nums)
        def backtrack(nums, used):
            if len(path) == len(nums):
                res.append(path[:])
                return
            prev = -11
            for i in range(len(nums)):
                if not used[i]:
                    if prev == -11 or prev != nums[i]:
                        prev = nums[i]
                        path.append(nums[i])
                        used[i] = True
                        backtrack(nums, used)
                        used[i] = False
                        path.pop()
        nums.sort()
        backtrack(nums, used)
        return res;

File: C++/47. Permutations II.cpp
/*
横向遍历不重复数字，纵向递归不重复数字
*/
class Solution {
public:
    vector<int> path;
    vector<vector<int>> res;
    void backtrack(vector<int>& nums, vector<bool>& used) {
        if (path.size() == nums.size()) {
            res.push_back(path);
            return;
        }
        int prev = -11;
        for (int i = 0; i < nums.size(); i++) {
            if (!used[i]) {
                if (prev == -11 || nums[i] != prev) {
                    prev = nums[i];
                    path.push_back(nums[i]);
                    used[i] = true;
                    backtrack(nums, used);
                    used[i] = false;
                    path.pop_back();
                }
            }

        }
    }
    vector<vector<int>> permuteUnique(vector<int>& nums) {
        path.clear();
        res.clear();
        vector<bool> used(nums.size(), false);
        sort(nums.begin(), nums.end());
        backtrack(nums, used);
        return res;
    }
};